helix-term/src/ui/picker.rs

```diff
@@ -30,11 +30,20 @@ type FileLocation = (PathBuf, Option<(usize, usize)>);
 pub struct FilePicker<T> {
     picker: Picker<T>,
     /// Caches paths to documents
-    preview_cache: HashMap<PathBuf, Document>,
+    preview_cache: HashMap<PathBuf, Preview>,
+    read_buffer: Vec<u8>,
     /// Given an item in the picker, return the file path and line number to display.
     file_fn: Box<dyn Fn(&Editor, &T) -> Option<FileLocation>>,
 }
 
+/// Caluculated file preview
+pub enum Preview {
+    Document(Document),
+    Binary,
+    LargeFile,
+    NotFound,
+}
```

Comment:
This is for my understanding, We need something like this?

```suggestion
/// Caluculated file preview
pub enum Preview<'a> {
    OwenedDocument(Document),
    BorrowedDocument(&'a Document),
    Binary,
    LargeFile,
    NotFound,
}
```

Reply:
Yeah pretty much. The tricky part is the lifetimes :/

Reply:
Thank you, that lifetime must be tricky. 
One of a easy way is add enum variant `FoundOnEditor` and render() check that. (This is not beautiful)
I'm looking forward to see your way.

Reply:
I arrived at a pretty similar solution, and then there was a `cannot borrow as immutable because also borrowed as mutable` error on L87 and L118 which I managed to solve using the entry API.